File: sdk-node/examples/entitlements/error-handling.ts
/**
 * Demonstrate the typed error tree on the entitlements surface. Each subclass
 * extends `ThreeCommonError`; branch with `instanceof`.
 *
 * Run:
 *   npx tsx examples/entitlements/error-handling.ts
 */

import {
  ThreeCommon,
  ThreeCommonAuthError,
  ThreeCommonConflictError,
  ThreeCommonNotFoundError,
  ThreeCommonRateLimitError,
  ThreeCommonValidationError,
} from '@3common/sdk'

const client = new ThreeCommon({
  apiKey: '3co_your_api_key_here',
})

try {
  await client.entitlements.consume({
    contactId: 'cnt_replace_with_real_id',
    featureKey: 'api_calls',
    amount: 1_000_000,
  })
} catch (err) {
  if (err instanceof ThreeCommonConflictError) {
    console.warn('insufficient balance — top up before consuming')
  } else if (err instanceof ThreeCommonNotFoundError) {
    console.warn('no entitlement record for this contact + feature')
  } else if (err instanceof ThreeCommonValidationError) {
    console.warn(`validation: ${err.message}`)
  } else if (err instanceof ThreeCommonAuthError) {
    console.warn('bad or expired API key')
  } else if (err instanceof ThreeCommonRateLimitError) {
    const wait = err.retryAfterSeconds ?? 30
    console.warn(`rate limited; retry in ${String(wait)}s`)
  } else {
    throw err
  }
}

File: sdk-python/examples/entitlements/error_handling.py
"""Run with: python examples/entitlements/error_handling.py"""

from __future__ import annotations

from threecommon import (
    AuthError,
    ConflictError,
    NotFoundError,
    RateLimitError,
    ThreeCommon,
)
from threecommon.entitlements import ConsumeBody


def main() -> None:
    with ThreeCommon(api_key="3co_your_api_key_here") as client:
        try:
            client.entitlements.consume(
                ConsumeBody(
                    contact_id="cnt_replace_with_real_id",
                    feature_key="api_calls",
                    amount=1_000_000,
                )
            )
        except ConflictError:
            print("insufficient balance — top up before consuming")
        except NotFoundError as e:
            print(f"no entitlement record for this contact + feature — request_id={e.request_id}")
        except AuthError as e:
            print(f"auth failed: check your API key — code={e.code}")
        except RateLimitError as e:
            wait = e.retry_after_seconds or 30
            print(f"rate limited; waiting {wait}s before retry")


if __name__ == "__main__":
    main()
